File: haskell/src/Fog/Mesh.hs
-- | FogMesh — multi-agent fog layer with propagation.
--
-- FogMesh is an immutable value — a map from agent ID to FogMap.
-- All operations are pure: pass current POSIX time where decay is needed.
--
-- Propagation rule: when agent A lifts fog on key X, all agents holding
-- 'InferredUnknown' on X have their confidence halved. The topological
-- inference that X is universally unknown no longer holds.
--
-- The 'darkCore' is what no internal propagation can touch — gaps present
-- in every agent. External signal required to clear them.
module Fog.Mesh
  ( FogMesh(..)
  , emptyMesh
  , registerAgent
  , propagateLift
  , totalFogVolume
  , systemFogVolume
  , systemEntropy
  , unionGaps
  , darkCore
  , seamMatrix
  , highestTensionSeam
  , meshSnapshot
  , MeshSnapshot(..)
  ) where

import Data.Map.Strict (Map)
import qualified Data.Map.Strict as Map
import Data.Set (Set)
import qualified Data.Set as Set
import Data.List (maximumBy)
import Data.Ord (comparing)
import Fog.Gap (GapKind(..), Gap(..), effectiveConfidence)
import Fog.Map (FogMap(..), fogKeys, fogVolume, fogEntropy)
import Fog.Seam (FogSeam, measure, tension)

-- | A mesh of agents' fog maps.
data FogMesh = FogMesh
  { meshAgents   :: Map String FogMap
  , meshHalfLife :: Double
  } deriving (Eq, Show)

-- | Empty mesh.
emptyMesh :: Double -> FogMesh
emptyMesh hl = FogMesh { meshAgents = Map.empty, meshHalfLife = hl }

-- | Register an agent's fog map.
registerAgent :: FogMap -> FogMesh -> FogMesh
registerAgent fmap mesh =
  mesh { meshAgents = Map.insert (fogAgentId fmap) fmap (meshAgents mesh) }
  where fogAgentId = \f -> case f of FogMap aid _ -> aid

-- | Propagate a fog lift from @agentId@ on @key@.
--
-- For every other agent holding 'InferredUnknown' on the same key,
-- multiply their confidence by @factor@ (typically 0.5).
-- Returns the updated mesh and the list of agent IDs that were updated.
propagateLift
  :: String      -- ^ Agent lifting fog
  -> String      -- ^ Gap key (already full key, or pass domain separately)
  -> Double      -- ^ Propagation factor (e.g. 0.5)
  -> FogMesh
  -> (FogMesh, [String])
propagateLift agentId key factor mesh =
  let (updatedAgents, touched) =
        Map.foldlWithKey' step (meshAgents mesh, []) (meshAgents mesh)
  in (mesh { meshAgents = updatedAgents }, reverse touched)
  where
    step (agents, acc) aid fmap
      | aid == agentId = (agents, acc)
      | otherwise =
          case Map.lookup key (fogGaps fmap) of
            Just gap | gapKind gap == InferredUnknown ->
              let newGap   = gap { gapConfidence = max (gapConfidence gap * factor) 0.0 }
                  newFmap  = fmap { fogGaps = Map.insert key newGap (fogGaps fmap) }
              in (Map.insert aid newFmap agents, aid : acc)
            _ -> (agents, acc)

-- | Sum of all effective confidences across all agents (not deduplicated).
totalFogVolume :: Double -> FogMesh -> Double
totalFogVolume now mesh =
  Map.foldl' (\acc fmap -> acc + fogVolume now (meshHalfLife mesh) fmap) 0.0 (meshAgents mesh)

-- | Deduplicated fog volume: each gap key counted once at max confidence.
systemFogVolume :: Double -> FogMesh -> Double
systemFogVolume now mesh =
  let hl  = meshHalfLife mesh
      maxByKey = Map.foldl'
        (\acc fmap ->
           Map.foldlWithKey' (\m k gap ->
             let e = effectiveConfidence now hl gap
             in Map.insertWith max k e m
           ) acc (fogGaps fmap)
        ) Map.empty (meshAgents mesh)
  in Map.foldl' (+) 0.0 maxByKey

-- | Shannon entropy of the total mesh fog distribution.
systemEntropy :: Double -> FogMesh -> Double
systemEntropy now mesh =
  let hl   = meshHalfLife mesh
      effs = [ effectiveConfidence now hl gap
             | fmap <- Map.elems (meshAgents mesh)
             , gap  <- Map.elems (fogGaps fmap) ]
      total = sum effs
  in if total == 0.0
     then 0.0
     else negate $ sum [ let p = e / total in if p > 0 then p * logBase 2 p else 0.0
                       | e <- effs ]

-- | All gap keys present in any agent.
unionGaps :: FogMesh -> Set String
unionGaps mesh =
  Map.foldl' (\acc fmap -> Set.union acc (fogKeys fmap)) Set.empty (meshAgents mesh)

-- | Gaps present in ALL agents — genuine system-wide blind spots.
-- External signal required to clear these.
darkCore :: FogMesh -> Set String
darkCore mesh =
  case Map.elems (meshAgents mesh) of
    []     -> Set.empty
    (f:fs) -> foldl (\acc fmap -> Set.intersection acc (fogKeys fmap)) (fogKeys f) fs

-- | All pairwise seams.
seamMatrix :: Double -> FogMesh -> Map (String, String) FogSeam
seamMatrix now mesh =
  let hl     = meshHalfLife mesh
      agents = Map.elems (meshAgents mesh)
      pairs  = [ (a, b) | (a:rest) <- tails agents, b <- rest ]
  in Map.fromList
       [ ((fogAgentIdOf a, fogAgentIdOf b), measure now hl a b) | (a, b) <- pairs ]
  where
    fogAgentIdOf (FogMap aid _) = aid
    tails []     = []
    tails (x:xs) = (x:xs) : tails xs

-- | Which pair of agents has the most to offer each other?
highestTensionSeam :: Double -> FogMesh -> Maybe FogSeam
highestTensionSeam now mesh =
  let matrix = seamMatrix now mesh
  in if Map.null matrix
     then Nothing
     else Just . snd . maximumBy (comparing (tension . snd)) . Map.toList $ matrix

-- | Summary of the current mesh state.
data MeshSnapshot = MeshSnapshot
  { snapshotAgents        :: Int
  , snapshotTotalVolume   :: Double
  , snapshotSystemVolume  :: Double
  , snapshotSystemEntropy :: Double
  , snapshotUnionGaps     :: Int
  , snapshotDarkCore      :: Int
  } deriving (Eq, Show)

meshSnapshot :: Double -> FogMesh -> MeshSnapshot
meshSnapshot now mesh = MeshSnapshot
  { snapshotAgents        = Map.size (meshAgents mesh)
  , snapshotTotalVolume   = roundAt3 (totalFogVolume  now mesh)
  , snapshotSystemVolume  = roundAt3 (systemFogVolume now mesh)
  , snapshotSystemEntropy = roundAt3 (systemEntropy   now mesh)
  , snapshotUnionGaps     = Set.size (unionGaps mesh)
  , snapshotDarkCore      = Set.size (darkCore  mesh)
  }
  where
    roundAt3 x = fromIntegral (round (x * 1000) :: Int) / 1000.0

File: elixir/lib/fog/mesh.ex
defmodule Fog.Mesh do
  @moduledoc """
  Multi-agent fog layer with propagation.

  FogMesh is an immutable value — a map of agent_id → FogMap.
  Operations return new meshes (or lists of updated meshes).

  Propagation rule: when agent A lifts fog on key X, all other agents
  holding :inferred_unknown on X have their confidence halved — the
  topological inference that X is universally unknown no longer holds.

  The dark_core is what no internal propagation can touch — gaps present
  in every agent's map. External signal required to clear these.

  To use as a stateful process, wrap in a GenServer:

      defmodule FogMeshServer do
        use GenServer
        def init(half_life),                do: {:ok, Fog.Mesh.new(half_life)}
        def handle_call({:register, fog},   _, mesh), do: {:reply, :ok, Fog.Mesh.register(mesh, fog)}
        def handle_call(:snapshot,          _, mesh), do: {:reply, Fog.Mesh.snapshot(mesh), mesh}
        def handle_call({:propagate, a, k}, _, mesh) ->
          {mesh2, updated} = Fog.Mesh.propagate_lift(mesh, a, k)
          {:reply, updated, mesh2}
        end
      end
  """

  alias Fog.{Map, Gap, Seam}

  @type t :: %__MODULE__{
    agents:    %{String.t() => Map.t()},
    half_life: non_neg_integer()
  }

  defstruct agents: %{}, half_life: 7 * 24 * 3600

  @propagation_factor 0.5

  @doc "Create an empty mesh."
  def new(half_life \\ 7 * 24 * 3600) do
    %__MODULE__{half_life: half_life}
  end

  @doc "Register an agent's fog map. Returns updated mesh."
  def register(%__MODULE__{} = mesh, %Map{} = fog_map) do
    %{mesh | agents: Elixir.Map.put(mesh.agents, fog_map.agent_id, fog_map)}
  end

  @doc """
  Propagate a fog lift from agent_id on key across the mesh.

  For every other agent holding :inferred_unknown on the same key,
  multiply their confidence by the propagation factor (default 0.5).

  Returns {updated_mesh, [agent_ids that were updated]}.
  """
  def propagate_lift(%__MODULE__{} = mesh, agent_id, key, domain \\ nil,
                     factor \\ @propagation_factor) do
    full_key = if domain, do: "#{domain}:#{key}", else: key

    {updated_agents, touched} =
      mesh.agents
      |> Enum.reduce({mesh.agents, []}, fn {aid, fmap}, {agents_acc, touched_acc} ->
           if aid == agent_id do
             {agents_acc, touched_acc}
           else
             case Elixir.Map.get(fmap.gaps, full_key) do
               %Gap{kind: :inferred_unknown} = gap ->
                 new_gap   = %{gap | confidence: max(gap.confidence * factor, 0.0),
                                     last_seen: System.system_time(:second)}
                 new_fmap  = %{fmap | gaps: Elixir.Map.put(fmap.gaps, full_key, new_gap)}
                 {Elixir.Map.put(agents_acc, aid, new_fmap), [aid | touched_acc]}
               _ ->
                 {agents_acc, touched_acc}
             end
           end
         end)

    {%{mesh | agents: updated_agents}, Enum.reverse(touched)}
  end

  @doc "Sum of all effective confidences across all agents (not deduplicated)."
  def total_fog_volume(%__MODULE__{} = mesh) do
    mesh.agents
    |> Elixir.Map.values()
    |> Enum.reduce(0.0, &(Fog.Map.fog_volume(&1, mesh.half_life) + &2))
  end

  @doc """
  Deduplicated fog volume: each gap key counted once at max confidence.
  Measures actual system ignorance, not agent-redundant ignorance.
  """
  def system_fog_volume(%__MODULE__{} = mesh) do
    mesh.agents
    |> Elixir.Map.values()
    |> Enum.reduce(%{}, fn fmap, acc ->
         Enum.reduce(fmap.gaps, acc, fn {k, gap}, inner_acc ->
           eff = Gap.effective_confidence(gap, mesh.half_life)
           Elixir.Map.update(inner_acc, k, eff, &max(&1, eff))
         end)
       end)
    |> Elixir.Map.values()
    |> Enum.sum()
  end

  @doc """
  Shannon entropy of the total mesh fog distribution.
  Treats all gaps across all agents as a single flat distribution.
  """
  def system_entropy(%__MODULE__{} = mesh) do
    effs =
      mesh.agents
      |> Elixir.Map.values()
      |> Enum.flat_map(fn fmap ->
           Enum.map(fmap.gaps, fn {_, gap} -> Gap.effective_confidence(gap, mesh.half_life) end)
         end)

    total = Enum.sum(effs)
    if total == 0.0 do
      0.0
    else
      Enum.reduce(effs, 0.0, fn e, acc ->
        p = e / total
        if p > 0, do: acc - p * :math.log2(p), else: acc
      end)
    end
  end

  @doc "All gap keys present in any agent."
  def union_gaps(%__MODULE__{} = mesh) do
    mesh.agents
    |> Elixir.Map.values()
    |> Enum.reduce(MapSet.new(), fn fmap, acc ->
         MapSet.union(acc, fmap.gaps |> Elixir.Map.keys() |> MapSet.new())
       end)
  end

  @doc """
  Gaps present in ALL agents — genuine system-wide blind spots.
  No internal propagation can clear these. External signal required.
  """
  def dark_core(%__MODULE__{} = mesh) do
    case Elixir.Map.values(mesh.agents) do
      []    -> MapSet.new()
      fmaps ->
        fmaps
        |> Enum.map(fn fmap -> fmap.gaps |> Elixir.Map.keys() |> MapSet.new() end)
        |> Enum.reduce(&MapSet.intersection/2)
    end
  end

  @doc "All pairwise seams across the mesh."
  def seam_matrix(%__MODULE__{} = mesh) do
    agents = Elixir.Map.values(mesh.agents)

    for {a, i} <- Enum.with_index(agents),
        b      <- Enum.drop(agents, i + 1),
        into: %{} do
      {{a.agent_id, b.agent_id}, Seam.measure(a, b, mesh.half_life)}
    end
  end

  @doc "Which pair of agents has the most to offer each other?"
  def highest_tension_seam(%__MODULE__{} = mesh) do
    matrix = seam_matrix(mesh)
    if map_size(matrix) == 0 do
      nil
    else
      {_pair, seam} = Enum.max_by(matrix, fn {_, s} -> Seam.tension(s) end)
      seam
    end
  end

  @doc "Summary metrics for the current mesh state."
  def snapshot(%__MODULE__{} = mesh) do
    %{
      agents:         map_size(mesh.agents),
      total_volume:   Float.round(total_fog_volume(mesh), 3),
      system_volume:  Float.round(system_fog_volume(mesh), 3),
      system_entropy: Float.round(system_entropy(mesh), 3),
      union_gaps:     MapSet.size(union_gaps(mesh)),
      dark_core:      MapSet.size(dark_core(mesh))
    }
  end
end
